From 62e0155c1aa7a7c9d66bf35ae46e79e2bbc148ce Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:25:02 -0400
Subject: [tools] perf/trace: Beautify flock 'cmd' arg

Message-id: <1408462094-14194-134-git-send-email-jolsa@redhat.com>
Patchwork-id: 88132
O-Subject: [PATCH RHEL7.1 BZ1131394 133/325] perf trace: Beautify flock 'cmd' arg
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 5cea6ff265e6979226b751f5ebfdda4594322523
Author: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Fri Sep 20 11:49:50 2013 -0300

description
===========
  4735.638 ( 0.003 ms): man/19881 flock(fd: 3, cmd: SH|NB) = 0
  4735.832 ( 0.002 ms): man/19881 flock(fd: 3, cmd: UN   ) = 0
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-trace.c b/tools/perf/builtin-trace.c
index c712a78..de37641 100644
--- a/tools/perf/builtin-trace.c
+++ b/tools/perf/builtin-trace.c
@@ -189,6 +189,37 @@ static size_t syscall_arg__scnprintf_madvise_behavior(char *bf, size_t size,
 
 #define SCA_MADV_BHV syscall_arg__scnprintf_madvise_behavior
 
+static size_t syscall_arg__scnprintf_flock(char *bf, size_t size,
+        struct syscall_arg *arg)
+{
+ int printed = 0, op = arg->val;
+
+ if (op == 0)
+  return scnprintf(bf, size, "NONE");
+#define P_CMD(cmd) \
+ if ((op & LOCK_##cmd) == LOCK_##cmd) { \
+  printed += scnprintf(bf + printed, size - printed, "%s%s", printed ? "|" : "", #cmd); \
+  op &= ~LOCK_##cmd; \
+ }
+
+ P_CMD(SH);
+ P_CMD(EX);
+ P_CMD(NB);
+ P_CMD(UN);
+ P_CMD(MAND);
+ P_CMD(RW);
+ P_CMD(READ);
+ P_CMD(WRITE);
+#undef P_OP
+
+ if (op)
+  printed += scnprintf(bf + printed, size - printed, "%s%#x", printed ? "|" : "", op);
+
+ return printed;
+}
+
+#define SCA_FLOCK syscall_arg__scnprintf_flock
+
 static size_t syscall_arg__scnprintf_futex_op(char *bf, size_t size, struct syscall_arg *arg)
 {
  enum syscall_futex_args {
@@ -554,6 +585,8 @@ static struct syscall_fmt {
  { .name     = "fcntl",     .errmsg = true,
    .arg_scnprintf = { [1] = SCA_STRARRAY, /* cmd */ },
    .arg_parm  = { [1] = &strarray__fcntl_cmds, /* cmd */ }, },
+ { .name     = "flock",     .errmsg = true,
+   .arg_scnprintf = { [1] = SCA_FLOCK, /* cmd */ }, },
  { .name     = "fstat",     .errmsg = true, .alias = "newfstat", },
  { .name     = "fstatat",    .errmsg = true, .alias = "newfstatat", },
  { .name     = "futex",     .errmsg = true,
-- 
1.7.1